go/dexop/internal/controller/client_controller.go

```diff
@@ -0,0 +1,236 @@
+	if deleteRequested {
+		if controllerutil.ContainsFinalizer(clientSpec, dexFinalizer) {
+			if _, err := r.DexManager.RemoveOauth2Client(clientSpec.Spec.Name); err != nil {
+				return deleteRequested, err
+			}
+
+			// remove finalizer
+			controllerutil.RemoveFinalizer(clientSpec, dexFinalizer)
+			err := r.Update(ctx, clientSpec)
+			if err != nil {
+				return deleteRequested, err
+			}
+		}
+		return deleteRequested, nil
+	}
+	return deleteRequested, nil
+}
+
+// adds finalizer on the Resource so that deletion can be handled gracefully
+func (r *ClientReconciler) addFinalizer(ctx context.Context, clientSpec *dexv1alpha1.Client) error {
+	// add finalizer
+	if !controllerutil.ContainsFinalizer(clientSpec, dexFinalizer) {
+		controllerutil.AddFinalizer(clientSpec, dexFinalizer)
+		err := r.Update(ctx, clientSpec)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Handle reading/generating Kubernetes secret to setup the password that will be used by Oauth2 client
+func (r *ClientReconciler) manageSecret(ctx context.Context, req ctrl.Request, clientSpec *dexv1alpha1.Client, reqLogger logr.Logger) error {
+	secretmgr := new(SecretManager)
+	if clientSpec.Spec.SecretName == "" {
+		return nil
+	}
+
+	if clientSpec.Spec.SecretNamespace == "" {
+		clientSpec.Spec.SecretNamespace = req.Namespace
+	}
+
+	secretValue, err := r.readOrGenerateSecret(ctx, secretmgr, clientSpec, reqLogger)
+	if err != nil {
+		return err
+	}
+
+	clientSpec.Spec.SecretValue = secretValue
+	return nil
+}
+
+// reads the password from an existing Secret, fallback to generating new one
+func (r *ClientReconciler) readOrGenerateSecret(ctx context.Context, secretmgr *SecretManager, clientSpec *dexv1alpha1.Client, reqLogger logr.Logger) (string, error) {
+	secret, err := secretmgr.readSecret(r, ctx, clientSpec.Spec.SecretName, clientSpec.Spec.SecretNamespace)
+	if err != nil {
+		if errors.IsNotFound(err) && clientSpec.Spec.GenerateSecret {
+			return r.generateSecret(ctx, secretmgr, clientSpec, reqLogger)
+		}
+		reqLogger.Error(err, "Unable to read secret", "secretName", clientSpec.Spec.SecretName)
+		return "", err
+	}
+
+	if secret == "" {
+		reqLogger.Error(nil, "Secret data is missing", "SecretName", clientSpec.Spec.SecretName)
+		return "", nil
+	}
+
+	return secret, nil
+}
+
+// generateSecret creates a Kubernetes secret with randomly generated password.
+func (r *ClientReconciler) generateSecret(ctx context.Context, secretmgr *SecretManager, clientSpec *dexv1alpha1.Client, reqLogger logr.Logger) (string, error) {
+	secret, err := secretmgr.generateSecret(r, ctx, clientSpec)
+	if err != nil {
+		reqLogger.Error(err, "Unable to write secret", "secretName", clientSpec.Spec.SecretName)
+		return "", err
+	}
+
+	if err = ctrl.SetControllerReference(clientSpec, secret, r.Scheme); err != nil {
+		return "", err
+	}
+
+	if err = r.Update(ctx, secret); err != nil {
+		return "", err
+	}
+
+	if secret.Data["secret"] == nil {
```

Comment:
This can cause panic if secret.Data map is nil or secret key is not present.

we can add additional check, in that way it will first check if map is not not nil
```go
if secret.Data == nil || secret.Data["secret"] == nil {
```

Reply:
good catch, fixed - thx!